[ios][android] Add fee indicator to Place Page and search results

dvdmrtnz opened the pull request:
This PR adds a fee indicator to the Place Page in iOS. This icon will show in any place that has a `fee=yes` property.

* Related to https://github.com/organicmaps/organicmaps/issues/3192

* Closes https://github.com/organicmaps/organicmaps/issues/2707

* Closes https://github.com/organicmaps/organicmaps/issues/5308

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/0dafef9b-4c29-4e93-bb2f-db4c5f2bc50c" width="310"/>
<img src="https://github.com/organicmaps/organicmaps/assets/47610359/f8f6adfb-049c-4f80-976c-2baf9d78de14" width="310"/>
<img src="https://github.com/organicmaps/organicmaps/assets/47610359/f41a8a47-d187-4096-89bb-a82f167ce08b" width="310"/>

Jean-BaptisteC commented on map/place_page_info.cpp:
```diff
@@ -162,6 +162,11 @@ std::string Info::FormatSubtitle(bool withType) const
+  auto const fee = GetMetadata(feature::Metadata::FMD_FEE);
+  if (!fee.empty())
+    if (fee == "yes")
+      append("💰");
```
It's not better to use [unicode number](https://symbl.cc/en/1F4B0/) like wheelchair icon?

dvdmrtnz replied:
Changed to text

matheusgomesms commented:
Great job!!! I always wanted that differentiation! Regarding the icon, I would personally use the dollar sign. I think it’s more recognizable than the bag one.

strump commented on map/place_page_info.cpp:
```diff
@@ -162,6 +162,11 @@ std::string Info::FormatSubtitle(bool withType) const
   if (GetWheelchairType() == ftraits::WheelchairAvailability::Yes)
     append(kWheelchairSymbol);
 
+  auto const fee = GetMetadata(feature::Metadata::FMD_FEE);
+  if (!fee.empty())
+    if (fee == "yes")
+      append("💰");
```
Plese move this logic to new method `bool HasFee()`. See how `bool HasWifi()` is implemented as an example.

dvdmrtnz replied:
Moved logic to GetLocalizedFee()

dvdmrtnz commented:
> Great job!!! I always wanted that differentiation! Regarding the icon, I would personally use the dollar sign. I think it’s more recognizable than the bag one.

🤔I’m thinking that maybe it could be better to have it in text: Fee/No fee. That way we could have info about the feature being free, as opposed to no info available. Because I don’t thing there is a suitable "No fee" icon. WDYT?

dvdmrtnz commented:
I've changed it to a localized Paid/Free text.



<img src="https://github.com/organicmaps/organicmaps/assets/47610359/6ff3f024-2e1c-4990-bc70-dbceb4d8ec22" width="350"/>
<img src="https://github.com/organicmaps/organicmaps/assets/47610359/fefd02ff-36c3-49d8-84ba-69d94b6a7f8d" width="350"/>

RedAuburn commented:
> I've changed it to a localized Paid/Free text.

this looks good :)

Jean-BaptisteC commented:
What about Android?

dvdmrtnz commented:
> What about Android?

I’m not sure but I think it should actually work in Android with these changes, since I’ve made this modification in the c++ core

Jean-BaptisteC commented:
I see nothing on Android with latest build

dvdmrtnz commented:
> I see nothing on Android with latest build

Maps need to be regenerated because there is a new metadata property

map-per reviewed (commented):
Added a suggestion for the german translation

dvdmrtnz commented:
> I see nothing on Android with latest build

I just tested it on Android. A minor change was needed, and now it works there too.

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/98b98698-8e90-441e-9cb4-1909bed5ff82" width="310"/>

biodranik commented:
Did you see https://github.com/organicmaps/organicmaps/issues/1312 and https://github.com/organicmaps/organicmaps/issues/2254 ?

How important is it to be able to _quickly_ search by the value of the fee tag?

dvdmrtnz commented:
> Did you see #1312 and #2254 ?
> 
> How important is it to be able to _quickly_ search by the value of the fee tag?

Yes, I saw those threads. I guess that you are asking me this because I took the "metadata" approach to adding this info, instead of adding it as a "type", right? I thought it was easier to add `fee=*` as metadata property and more scalable (given the limit of  7 types per feature), specially if we want to add some more properties like [`takeaway=*`](https://wiki.openstreetmap.org/wiki/Key:takeaway), [`delivery=*`](https://wiki.openstreetmap.org/wiki/Key:delivery), [`capacity=*`](https://wiki.openstreetmap.org/wiki/Key:capacity), [`drive_through=*`](https://wiki.openstreetmap.org/wiki/Key:drive_through) and similar properties of other bigger places.

biodranik commented:
fee is important for some other types too like parkings, toilets, etc. Using metadata in the search is possible, but it will be _very_ slow.

vng reviewed (commented):
UI - no problem here.

I'm planning to refactor fee, access, etc https://github.com/organicmaps/organicmaps/issues/1312

Storing fee in metadata is not a good idea. At least because of fast reading for rendering. I'm sure, we will render some dynamic icons someday depending on these tags.

dvdmrtnz commented:
> UI - no problem here.
> 
> I'm planning to refactor fee, access, etc #1312
> 
> Storing fee in metadata is not a good idea. At least because of fast reading for rendering. I'm sure, we will render some dynamic icons someday depending on these tags.

Okay. I guess we'll have to wait until the fee refactoring then.

pastk commented:
> Storing fee in metadata is not a good idea. At least because of fast reading for rendering. I'm sure, we will render some dynamic icons someday depending on these tags.

I think this solution is fine for the interim.

The fee/access refactoring mentioned is a very long shot. And dynamic icons is even longer one :) No need to wait for them when we can improve UX now.

dvdmrtnz commented:
> We've been discussing it and decided that adding separate `fee-yes`/`fee-no` types seems to be a better long-term approach.

I have tried adding to `mapcss-mapping.csv`:

```
fee|yes;[fee=yes];;name;int_name;1583;
fee|no;[fee=no];;name;int_name;1584;
```

But it doesn't seem to be working 🤔. I have regenerated maps and everything but it doesn't show up in the raw types. Any idea why?

https://www.openstreetmap.org/node/9341902098#map=19/42.81552/-1.64178
![](https://github.com/organicmaps/organicmaps/assets/47610359/29c3797a-8c9f-48a6-aa13-1a3afa9fda77)

pastk commented:
> But it doesn't seem to be working 🤔. I have regenerated maps and everything but it doesn't show up in the raw types. Any idea why?

Types that don't have any styles defined are discarded by the generator.
So you've got to add these new types into "exceptions" alongside `cuisine` and `wheelchair` in `indexer/feature_visibility.cpp::IsUsefulNondrawableType()`.

Please add a note about this behavior into the comment in the `mapcss-mapping.csv`.

Also note that its better to use `fee|yes;[fee?];...` as it'll include non-yes values like `fee=10` (basically everything except `fee=no`).

dvdmrtnz commented:
@pastk @vng @biodranik PTAL. Now using fee-yes, fee-no types, instead of metadata

pastk commented on data/strings/types_strings.txt:
```diff
@@ -10910,6 +10910,85 @@
     zh-Hant = 出口
 
+  [type.fee.yes]
+    en = Paid
```
so for an `amenity=parking fee=yes` feature it'll be `Paid Parking • Paid` in the PP subtitle, which doesn't look very good...

@vng does search use type's translation for searching? i.e. does presence of "Paid" in the translation make a difference?
if not, then we can probably drop it and leave just "Parking"

dvdmrtnz replied:
AFAIK there are no parking-fee types in categories so currently you cannot search for "Paid Parking" exactly. However when you search for "Parking" you will see the description as "Paid Parking" if that's the case.

dvdmrtnz replied:
I think we should add the "Paid"/"Free" indicator to search results too and simplify the "Paid Parking" strings to "Parking" so it would read as "Parking • Paid" or "Parking • Free" both in the PlacePage and in the search results

dvdmrtnz commented:
* I have simplified Paid Parking strings so they don't repeat "Paid Parking • Paid", instead showing as "Parking • Paid"
* I have added the "Paid"/"Free" indicator to search results too, in iOS and Android

Here are some iOS/Android screenshots:

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/f119ea0c-99e2-4df8-aa35-dc72a88b76e7" width="320"/> <img src="https://github.com/organicmaps/organicmaps/assets/47610359/ae78d7f5-81d3-44a3-a50e-8b7c8320896b" width="320"/>

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/8da985c4-97e2-4e57-b7ee-22e9d5875959" width="320"/> <img src="https://github.com/organicmaps/organicmaps/assets/47610359/4140f497-1e05-4f9a-aaf8-2598e8d2ca95" width="320"/>

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/e82b5aa5-b97a-4754-b072-e80ce16225a9" width="320"/> <img src="https://github.com/organicmaps/organicmaps/assets/47610359/472b1a06-62f7-4915-815c-cc8ab9c7d328" width="320"/>

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/de2e22f5-7c1a-4924-80de-8b3ecc1372f4" width="320"/> <img src="https://github.com/organicmaps/organicmaps/assets/47610359/9f6d2967-b419-42fc-a5e0-7b0e2d5d2478" width="320"/>

dvdmrtnz commented on map/place_page_info.cpp:
```diff
@@ -162,6 +162,10 @@ std::string Info::FormatSubtitle(bool withType) const
     append(kWheelchairSymbol);
 
+  // Fee.
+  for (auto const & fee : GetLocalizedFeeTypes())
```
Done

dvdmrtnz commented on generator/generator_tests/osm_type_test.cpp:
```diff
@@ -1209,9 +1209,10 @@ UNIT_CLASS_TEST(TestWithClassificator, OsmType_ReuseTags)
+    TEST_EQUAL(params.m_types.size(), 3, (params));
     TEST(params.IsTypeExist(GetType({"amenity", "parking", "private"})), (params));
     TEST(params.IsTypeExist(GetType({"amenity", "parking", "fee"})), (params));
+    TEST(params.IsTypeExist(GetType({"fee", "yes"})), (params));
```
Done

Misalf-git commented:
Is it necessary / beneficial to to show "free" in the place page?

IMO it's easier to understand information based on whether a value is present or not, instead of interpreting it.

In german, "paid" and "free" might translate to something rather similar on a quick glimpse:
"Kostenpflichtig" and "Kostenfrei".
Also, in case of cut off text (maybe for more languages that tend to inflate character count) this could lead to something like
"Parkplatz · ​♿︀​ · Kost...".

dvdmrtnz commented:
> Is it necessary / beneficial to to show "free" in the place page?
> 
> IMO it's easier to understand information based on whether a value is present or not, instead of interpreting it.
> 
> In german, "paid" and "free" might translate to something rather similar on a quick glimpse: "Kostenpflichtig" and "Kostenfrei". 

@Misalf-git I think it's indeed beneficial to show "Free". You want to be able to distinguish no info vs no fee.

> Also, in case of cut off text (maybe for more languages that tend to inflate character count) this could lead to something like "Parkplatz · ​♿︀​ · Kost...".

That's no problem. The line is set to wrap around into multiple lines if it doesn't fit in one line. It's the same line as the cuisine line for example.

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/f30fcead-dc26-48f5-9f85-f91224fbe6c0" width="320"/> <img src="https://github.com/organicmaps/organicmaps/assets/47610359/fcfae47d-c605-40f6-a05b-05e6e266aa90" width="320"/>

Misalf-git commented:
OK. Then it's up to the translation to prevent excessive brain activity.

pastk commented on android/app/src/main/res/values-ru/strings.xml:
```diff
@@ -541,8 +541,6 @@
 	<string name="editor_share_to_all_dialog_message_2">Редакторы OpenStreetMap проверят изменения и свяжутся с вами, если у них возникнут вопросы.</string>
 	<string name="navigation_stop_button">стоп</string>
-	<!-- Text under the version number in the Help dialog. TODO: Synchronize other translations with English. -->
-	<string name="about_description">Organic Maps — быстрые и бесплатные карты, которые работают без Интернета. Все картографические данные берутся из OpenStreetMap.org, там можно самостоятельно исправлять ошибки и добавлять новые объекты. В Organic Maps нет рекламы и сбора персональных данных. Это проект с открытым исходным кодом, создаваемый энтузиастами в свободное время. Будем рады вашей поддерж ...
```
hm, why is it removed in this PR?
perhaps it was removed elsewhere but strings were not regenerated?

dvdmrtnz replied:
Must have been in some other PR. I didn't touch that string

pastk commented on search/intermediate_result.cpp:
```diff
@@ -291,6 +291,9 @@ void FillDetails(FeatureType & ft, Result::Details & details)
   auto const roadShields = ftypes::GetRoadShieldsNames(ft);
   details.m_roadShields = strings::JoinStrings(roadShields, osm::MapObject::kFieldsSeparator);
+    
+  auto const fee = feature::GetLocalizedFeeType(feature::TypesHolder(ft));
```
nit: would be good to init the TypesHolder once only and then re-use for all checks like GetLocalizedCuisines, IsHotelChecker...
(`IsHotelChecker::Instance()(ft);` inits a TypesHolder under the hood)

dvdmrtnz replied:
Done

biodranik commented:
1. With parking, is it better to display "Paid underground parking" than "Underground parking * paid"?
2. "Paid" word alone looks weird in some contexts (the opposite of "not paid" instead of "free"). That's why it would be great to think about some short but clear symbol (or letter?) instead of text.
3. @vng @pastk we have a limit of 7 types per feature, right? With this (and other added types), how many other types are lost? How important are these "lost types" cases?

pastk commented:
> 1. With parking, is it better to display "Paid underground parking" than "Underground parking * paid"?

Maybe. But this approach is not very scalable and will complicate the code unnecessary (probably we'll have to hardcode parkings as exceptions).
E.g. later we'd want to add "Private" also.

>     2. "Paid" word alone looks weird in some contexts (the opposite of "not paid" instead of "free"). That's why it would be great to think about some short but clear symbol (or letter?) instead of text.
> 

https://github.com/organicmaps/organicmaps/pull/5883#issuecomment-1702449753

>     3. @vng @pastk we have a limit of 7 types per feature, right? With this (and other added types), how many other types are lost? How important are these "lost types" cases?

It happens mostly with cuisines and recycling types.
(and the limit is **8** types actually)

> 
>     4. @vng you've mentioned types refactoring, how is it related to this change?

Its first PR that converts a previously type-embedded attribute (fee) into a separate type
- #1312

pastk commented:
> 2\. "Paid" word alone looks weird in some contexts (the opposite of "not paid" instead of "free"). That's why it would be great to think about some short but clear symbol (or letter?) instead of text.

But actaully it might look OK if we use a symbol/icon for `fee-yes` and a `Free` text for `fee-no`..

pastk commented on search/intermediate_result.cpp:
```diff
@@ -279,20 +279,22 @@ void FillDetails(FeatureType & ft, Result::Details & details)
-
-  details.m_isHotel = ftypes::IsHotelChecker::Instance()(ft);
+  
+  auto types_holder = feature::TypesHolder(ft);
```
we use camelCase for variable names
`typesHolder`

dvdmrtnz replied:
Fixed

pastk commented:
> But actually it might look OK if we use a symbol/icon for `fee-yes` and a `Free` text for `fee-no`..

@vng Will "paid" and "free" become searchable out-of-the-box after this change (there is a type and a corresponding translation).
If so then using an emoji as a translation will affect searchability?

biodranik commented:
> If so then using an emoji as a translation will affect searchability?

No. data/categories.txt supports emoji as search synonyms.

dvdmrtnz commented:
Okay let's explore doing emojis. What about 💲for `fee-yes` and 💲✖ for `fee-no`:

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/ef3abf59-7932-4954-adf6-f9f7bc35601d" width="320"/> <img src="https://github.com/organicmaps/organicmaps/assets/47610359/b13e0955-a35b-4f2b-b59e-4f0a3ce0f4ff" width="320"/>

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/06fdf468-d8d0-458c-91c7-a386acde3cae" width="320"/> <img src="https://github.com/organicmaps/organicmaps/assets/47610359/b25d787f-00dc-4088-bfec-6c73ec3ae5ca" width="320"/>

dvdmrtnz commented:
Emojis make it definitely easier to quickly see at a glance which results are paid/free when searching:

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/83282733-0934-4fbc-bf26-345c30613ab6" width="320"/>

dvdmrtnz commented:
Also this cross emoji approach could be used in the future for indicating for example `wheelchair=no` (♿️✖) which is currently not shown as far as I know

Misalf-git commented:
The dollar sign may be recognizable by many, but IMO something international / neutral would be better.

I guess the generic "currency sign" is not known that much?
 ¤

And the "coin" emoji is probably too recent for older OSes.
 🪙

There are several currecny related symbols in unicode though.
 ₣₦₧₩₪€₭₮₯₱₲₳₴₵₸₹₺₼₽₾₿

Two separate icons to depict one thing could be misinterpreted. Maybe grouping might help.
 [♿❌]
Don't know if some can be combined in unicode.

pastk commented:
The "paid" dollar sign emoji is great, for me its very recognisable. Not sure if its universal though... I don't know any better alternative.
But the "free" version with added cross emoji is not so recognizable IMHO... And no better alternatives come to my mind..
I think its better to keep it as a text.

biodranik commented:
fee=no / fee=yes / no fee tag

(yes, it requires a bit more complex string processing, but it may be worth it)

1. Free Toilet / Paid Toilet / Toilet
2. Free Toilet / 💲 Toilet / Toilet
3. Toilet / 💲 Toilet / Toilet (people care if it's paid).
4. ~💲~ Toilet / 💲 Toilet / Toilet

dvdmrtnz commented:
> fee=no / fee=yes / no fee tag
> 
> (yes, it requires a bit more complex string processing, but it may be worth it)
> 
> 1. Free Toilet / Paid Toilet / Toilet
> 2. Free Toilet / 💲 Toilet / Toilet
> 3. Toilet / 💲 Toilet / Toilet (people care if it's paid).
> 4. ~💲~ Toilet / 💲 Toilet / Toilet

Please no string handling. It can get quite messy with all the languages. 

My options:

5. Toilet • Free / Toilet • Paid / Toilet
6. Toilet • 💲✖ / Toilet • 💲 / Toilet
7. Toilet • Free / Toilet • 💲 / Toilet

pastk commented:
Is there a crossed version of the 💲 emoji? Or a striked one like in biodranik's post?

Otherwise I like options 5 & 7 suggested by @dvdmrtnz 

Note that we already have a well-established system/notion of displaying POI's additional info via a ` • ` separator. I think we should use it for the fee info too for
1. consistency
2. to avoid complexity and high maintenance of custom strings like "Free Underground Parking", "Paid Private Toilet" etc.

Misalf-git commented:
Seems to me like a double-negative but Iet's see if any of those will get displayed properly...

0024+043F+20E0:
 $͏⃠ 
0024+FEFF+20E0:
 $﻿⃠
0024+2060+20E0:
 $⁠⃠ 
1F4B2+2060+20E0:
💲⁠⃠ 
1F10F:
 🄏

I'd go for option 5. or 7. too, though.

dvdmrtnz commented:
> Seems to me like a double-negative but Iet's see if any of those will get displayed properly...
> 
> 0024+043F+20E0: $͏⃠ 0024+FEFF+20E0: $﻿⃠ 0024+2060+20E0: $⁠⃠ 1F4B2+2060+20E0: 💲⁠⃠ 1F10F: 🄏
> 
> I'd go for option 5. or 7. too, though.

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/4056ee1a-222b-494c-af29-2cc8ab378618" width="320"/>

pastk commented:
IMHO none of those look good enough... Also their display might be dependent on available fonts, isn't it?

It might look OK if we make our own clean looking icon with a crossed dollar sign. But at the expense of hardcoding its display. And I don't think its worth it, a textual "Free" is a better option IMO.

dvdmrtnz commented:
Some more options:

\u0024\u20dd
$⃝

\u0024\u20e3
$⃣

\u0024\u20e5
$⃥

\u0024\u20ea
$⃪

\u0024\u20eb
$⃫

\u0024\u0336
$̶

\u0024\u20dd\u20e5
$⃝⃥

\u0024\u20dd\u0336
$⃝̶

\u0024\u20dd\u20eb
$⃝⃫

It's quite funny what you can do with unicode and diacritics 🤣. Here's the list of unicode diacritics: https://en.wikipedia.org/wiki/Diacritic#List_of_diacritics_in_Unicode

By the way I am using this website to try out the characters: https://www.branah.com/unicode-converter

Some of these look okay in iOS but then look quite bad on Android 😂

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/27b90900-55be-4f7f-8d04-5f65b536b1b0" width="320"/>

<img src="https://github.com/organicmaps/organicmaps/assets/47610359/d79f0348-1cf6-4570-bcf7-b5008974149b" width="320"/>

dvdmrtnz commented:
I'm leaving it as Toilet • Free / Toilet • 💲 / Toilet

biodranik commented:
@vng PTAL

vng commented on generator/generator_tests/osm_type_test.cpp:
```diff
@@ -2644,4 +2641,29 @@ UNIT_CLASS_TEST(TestWithClassificator, OsmType_ComplexTypesSmoke)
+
+  for (auto const & type : complexTypes)
+  {
+    auto const results = type.first;
```
auto const & results = ..

dvdmrtnz replied:
Done

vng commented on generator/generator_tests/osm_type_test.cpp:
```diff
@@ -2644,4 +2641,29 @@ UNIT_CLASS_TEST(TestWithClassificator, OsmType_ComplexTypesSmoke)
+    auto const results = type.first;
+    auto const params = GetFeatureBuilderParams(type.second);
+    TEST_EQUAL(params.m_types.size(), results.size(), (type, params));
+    for (auto const result : results)
```
auto const & result : ..

dvdmrtnz replied:
Done

vng commented on data/mapcss-mapping.csv:
```diff
@@ -1671,3 +1671,5 @@ amenity|parking|underground|private;[amenity=parking][location=underground][acce
 amenity|parking|lane|private;[amenity=parking][parking=lane][access=private];;name;int_name;1591;
 amenity|compressed_air;1592;
+fee|yes;[fee?];;name;int_name;1593;
+fee|no;[fee=no];;name;int_name;1594;
```
Let's move them up, somewhere near 500 instead of current deprecated?
I'd prefer to put _very_ common used types with small values.

dvdmrtnz replied:
Moved to 500 and 501

pastk replied:
A note for the future: regeneration is needed after `mapcss-mapping.csv` changes (atm the types order in `data/types.txt` is wrong on master).

vng reviewed (commented):
LGTM

vng commented on search/intermediate_result.cpp:
```diff
@@ -279,18 +279,23 @@ void FillDetails(FeatureType & ft, Result::Details & details)
-
-  details.m_isHotel = ftypes::IsHotelChecker::Instance()(ft);
+  
+  auto typesHolder = feature::TypesHolder(ft);
```
nit: feature::TypesHolder const typesHolder(ft);

dvdmrtnz replied:
Done

vng replied:
No, this is a complete expression, no need assign:
```
feature::TypesHolder const typesHolder(ft);
```

dvdmrtnz replied:
Fixed now

vng commented on search/intermediate_result.cpp:
```diff
@@ -279,18 +279,23 @@ void FillDetails(FeatureType & ft, Result::Details & details)
   details.m_roadShields = strings::JoinStrings(roadShields, osm::MapObject::kFieldsSeparator);
+    
+  auto const fee = feature::GetLocalizedFeeType(typesHolder);
+  details.m_fee = fee;
```
nit: details.m_fee = feature::GetLocalizedFeeType(typesHolder);
